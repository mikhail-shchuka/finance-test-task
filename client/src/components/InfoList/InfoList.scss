.list {
  margin: 50px;
  overflow-x: scroll;

  table {
    width: 100%;
    border: none;
    margin-bottom: 10px;
    border-collapse: separate;
    border-spacing: 0;
  
    thead th {
      font-weight: bold;
      text-align: center;
      border: none;
      padding: 20px 10px;
      background: #EDEDED;
      font-size: 20px;
      border-top: 1px solid #ddd;
    }
    
    tr th:first-child, .table tr td:first-child {
      border-left: 1px solid #ddd;
    }
    
    tr th:last-child, .table tr td:last-child {
      border-right: 1px solid #ddd;
    }
    
    thead tr th:first-child {
      border-radius: 20px 0 0 0;
    }
    
    thead tr th:last-child {
      border-radius: 0 20px 0 0;
    }
    
    tbody td {
      text-align: center;
      border: none;
      padding: 30px 10px;
      font-size: 20px;

      &:first-child {
        font-weight: bold;
      }

      &:last-child {
        padding: 10px;
      }

      div {
        width: max-content;
        margin: 0 auto;
      }
    }
    
    tbody tr:nth-child(even) {
      background: #F8F8F8;
    }
    
    tbody tr:last-child td{
      border-bottom: 1px solid #ddd;
    }
    
    tbody tr:last-child td:first-child {
      border-radius: 0 0 0 20px;
    }
    
    tbody tr:last-child td:last-child {
      border-radius: 0 0 20px 0;
    }
  }
}

.price {
  color: #137333;
  font-weight: bold;

  &--decrease {
    color: #EA3943;
  }
}
