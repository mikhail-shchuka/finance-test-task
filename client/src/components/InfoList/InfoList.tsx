import classNames from 'classnames';
import { Graph } from '../Graph';

import './InfoList.scss';
import { useTickersData } from '../../hooks/useTickersData';
import { getSignPrice } from '../../utils/getSignPrice';

export const InfoList = () => {
  const { prevTickers, tickers } = useTickersData();

  return (
    <div className="list">
      <table>
        <thead>
          <tr>
            <th>Ticker</th>
            <th>Exchange</th>
            <th>Price</th>
            <th>Change</th>
            <th>Change percent</th>
            <th>Dividend</th>
            <th>Yield</th>
            <th>Graph</th>
          </tr>
        </thead>
        <tbody>
          {tickers.map(ticker => {
            const sign = getSignPrice(prevTickers, ticker);

            return (
              <tr key={ticker.ticker}>
                <td>{ticker.ticker}</td>
                <td>
                  {ticker.exchange}
                </td>
                <td className={classNames(
                  'price',
                  { 'price--decrease': sign < 0 },
                )}
                >
                  {ticker.price}
                </td>
                <td>
                  <div>
                    {sign < 0 && '- '}
                    {ticker.change}
                  </div>
                </td>
                <td>{ticker.change_percent}</td>
                <td>{ticker.dividend}</td>
                <td>{ticker.yield}</td>
                <td>
                  <Graph tickerName={ticker.ticker} sign={sign} />
                </td>
              </tr>
            );
          })}
        </tbody>
      </table>
    </div>
  );
};
