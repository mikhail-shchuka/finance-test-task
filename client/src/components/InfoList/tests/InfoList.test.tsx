import { render, screen } from '@testing-library/react';
import { InfoList } from '../InfoList';
import { mockPrevTicker, mockTicker } from './constants';

const mockGraphTestId = 'Graph-test-id';

jest.mock('../../../hooks/useTickersData', () => ({
  useTickersData: () => (
    { tickers: [mockTicker], prevTickers: [mockPrevTicker] }),
}));

jest.mock('../../Graph', () => {
  return {
    Graph: () => (<div data-testid={mockGraphTestId} />),
  };
});

describe('InfoList', () => {
  it('should match with snapshot', () => {
    const { baseElement } = render(<InfoList />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should be rendered ticker info', () => {
    render(<InfoList />);

    expect(screen.getByTestId(mockGraphTestId)).toBeInTheDocument();
    expect(screen.getByText(mockTicker.ticker)).toBeInTheDocument();
  });
});
