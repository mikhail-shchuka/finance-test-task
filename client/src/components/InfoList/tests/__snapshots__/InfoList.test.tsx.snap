// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoList should match with snapshot 1`] = `
<body>
  <div>
    <div
      class="list"
    >
      <table>
        <thead>
          <tr>
            <th>
              Ticker
            </th>
            <th>
              Exchange
            </th>
            <th>
              Price
            </th>
            <th>
              Change
            </th>
            <th>
              Change percent
            </th>
            <th>
              Dividend
            </th>
            <th>
              Yield
            </th>
            <th>
              Graph
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              ticker
            </td>
            <td>
              exchange
            </td>
            <td
              class="price price--decrease"
            >
              100
            </td>
            <td>
              <div>
                - 
                change
              </div>
            </td>
            <td>
              change_percent
            </td>
            <td>
              dividend
            </td>
            <td>
              yield
            </td>
            <td>
              <div
                data-testid="Graph-test-id"
              />
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</body>
`;
